Script/AtomicEditor/ui/resourceEditors/AbstractTextResourceEditorBuilder.ts

```diff
@@ -0,0 +1,95 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+import EditorEvents = require("../../editor/EditorEvents");
+export abstract class AbstractTextResourceEditorBuilder implements Editor.Extensions.ResourceEditorBuilder {
+
+    abstract canHandleResource(resourcePath: string): boolean;
+
+    /**
+     * Full path is the fully qualified path from the root of the filesystem.  In order to take advantage
+     * of the resource caching system, let's trim it down to just the path inside the resources directory
+     * including the Resources directory so that the casing is correct
+     */
+    private getNormalizedPath(path: string) {
+        const RESOURCES_MARKER = "resources/";
+        return path.substring(path.toLowerCase().indexOf(RESOURCES_MARKER));
+    }
+
+    getEditor(resourceFrame: Atomic.UIWidget, resourcePath: string, tabContainer: Atomic.UITabContainer): Editor.ResourceEditor {
+
+        const editor = new Editor.JSResourceEditor(resourcePath, tabContainer);
+
+        // one time subscriptions waiting for the web view to finish loading.  This event
+        // actually hits the editor instance before we can hook it, so listen to it on the
+        // frame and then unhook it
+        editor.subscribeToEvent(EditorEvents.WebViewLoadEnd, (data) => {
+            editor.unsubscribeFromEvent(EditorEvents.WebViewLoadEnd);
+            this.loadCode(<Editor.JSResourceEditor>editor, resourcePath);
+        });
+
+        // Cannot subscribe to WebMessage until the .ts side can return a Handler.Success() message
+        // editor.subscribeToEvent(EditorEvents.WebMessage, (data) => {
+        //     console.log("editor Got a web message: " + data.request);
+        //     for (let x in data) {
+        //         console.log(x + ":" + data[x]);
+        //     }
+        //     const request = JSON.parse(data.request);
+        //     switch (request.message) {
+        //         case "editorGetUserPrefs":
+        //             this.getUserPrefs(editor);
+        //             return true;
+        //     }
+        // });
```

Comment:
This is more like reference code because once the WebMessage event can be handled in the typescript, this should be uncommented.